### PR TITLE
Add Postdominator tree

### PR DESCRIPTION
**Pull request by guillep**
Like the dominator but a block post dominates another one if it is a mandatory milestone towards the exit.
Thus this introduces an exit basic block that is mandatary in the CFG.

For now, all tests seem to pass as the post-dominator is not in use yet.
Probably will need some adjustments when we put this to work.

### CODE REVIEW

File: Druid-Tests/DruidTestRTLCompiler.class.st
Comment (PalumboN) [inline Druid-Tests/DruidTestRTLCompiler.class.st:1]: Why so many changes here? Is it sorting the basic blocks differently?

File: Druid/DRBasicBlock.class.st [hunk 1]
```diff
@@ -318,10 +318,18 @@ DRBasicBlock >> divide: aValue by: anotherValue [
 	^ self addInstruction: (self instructionFactory divide: aValue by: anotherValue)
 ]
 
+{ #category : #dominance }
+DRBasicBlock >> dominatedBlocks [
+	"Return true if the blocks in the argument are the only dominated blocks of the receiver.
+	This checks transitive domination"
```
Comment (PalumboN) [inline Druid/DRBasicBlock.class.st:324]: This comment doesn't apply.

```suggestion
```

File: Druid/DRBasicBlock.class.st [hunk 2]
```diff
@@ -499,6 +507,22 @@ DRBasicBlock >> id: anObject [
 	id := anObject
 ]
 
+{ #category : #dominance }
+DRBasicBlock >> immediatelyDominatedBlocks [
+	"Return true if the blocks in the argument are the only dominated blocks of the receiver.
+	This checks transitive domination"
```
Comment (PalumboN) [inline Druid/DRBasicBlock.class.st:513]:
```suggestion
```

File: Druid/DRBasicBlock.class.st [hunk 3]
```diff
@@ -499,6 +507,22 @@ DRBasicBlock >> id: anObject [
...
+
+	^ (controlFlowGraph dominatorTree childrenOf: self) asSet
+]
+
+{ #category : #dominance }
+DRBasicBlock >> immediatelyPostDominatedBlocks [
+	"Return true if the blocks in the argument are the only dominated blocks of the receiver.
+	This checks transitive domination"
```
Comment (PalumboN) [inline Druid/DRBasicBlock.class.st:521]:
```suggestion
```

File: Druid/DRBasicBlock.class.st [hunk 4]
```diff
@@ -647,28 +674,30 @@ DRBasicBlock >> intToFloat: aValue [
...
 	
-	^ false
+	^ basicBlocks allSatisfy: [ :b | b isDominatedBy: self ]
+]
+
+{ #category : #dominance }
+DRBasicBlock >> isDominatorOfAllOf: basicBlocks [
+	
+	^ basicBlocks allSatisfy: [ :b | b isDominatedBy: self ]
+]
```
Comment (PalumboN) [inline Druid/DRBasicBlock.class.st:693]: Duplicated methods

File: Druid/DRCogitSendMarshallHoisting.class.st [hunk 5]
```diff
@@ -21,8 +21,8 @@ DRCogitSendMarshallHoisting >> applyTo: cfg [
 	toMerge := cfg instructions select: [ :e | e isPop or: [ e isCogitSendMarshall ] ].
 	toMerge size <= 1 ifTrue: [ ^ self "nothing to do" ].
 
-	cfg buildDominatorTree.
-	dominator := cfg tree dominatorOfAll: toMerge.
+	cfg invalidateDominatorTree.
```
Comment (PalumboN) [inline Druid/DRCogitSendMarshallHoisting.class.st:24]: Niceeee 🤩 

We should call automatically this method on any change of predecessors / successors on any block.

Then, we don't need to force it any time that the dominator tree _could_ be used.

File: Druid/DRControlFlowGraph.class.st [hunk 6]
```diff
@@ -477,17 +507,18 @@ DRControlFlowGraph >> initialize [
...
 	edgesBySource := Dictionary new.
-	
+
 	"How many spill slots we need in memory to execute the compiled method"
 	numberOfSpillSlots := 0.
-	
-	nextBasicBlockId := 0.
-	initialBasicBlock := self privateNewBasicBlock.
+
+	nextBasicBlockId := -1.
```
Comment (PalumboN) [inline Druid/DRControlFlowGraph.class.st:519]: 😸

File: Druid/DRDominatorTree.class.st [hunk 7]
```diff
@@ -148,42 +188,81 @@ DRDominatorTree >> dominatorOfAll: instructions [
...
+			blockDictionary at: b ifAbsent: [
+				node := DRDominatorTreeNode new
+					        block: b;
+					        parent: currentNode;
+					        yourself.
+				currentNode addChild: node.
+
+				blockDictionary at: b put: node.
+				workList addFirst: node ] ] ]
+]
```
Comment (PalumboN) [inline Druid/DRDominatorTree.class.st:240]: This method is used? Could be defined based on 
`DRDominatorTree >> initialize: cfg startingFrom: initialBasicBlock iterating: aBlockClosure`

**Review by PalumboN — DISMISSED** [review]
Cooooolllll!!!! 🚀 

I left some comments, most of them just cleanings.

File: Druid/DRCogitSendMarshallHoisting.class.st (hunk 5 above)
Comment (PalumboN) [reply Druid/DRCogitSendMarshallHoisting.class.st:24]: https://github.com/Alamvic/druid/issues/140 :octocat:

File: Druid/DRBasicBlock.class.st (hunk 1 above)
Comment (guillep) [reply Druid/DRBasicBlock.class.st:324]: True, but the transitivity still applies here. I'll change the comment :)

Comment (guillep) [inline Druid/DRBasicBlock.class.st:324]:
```suggestion
	"Return a collection of all the blocks that are dominated (transitively) by the block given as argument"
```

File: Druid/DRBasicBlock.class.st (hunk 2 above)
Comment (guillep) [inline Druid/DRBasicBlock.class.st:513]:
```suggestion
	"Return a collection of all the blocks that are immediately dominated by the block given as argument"
```

Comment (guillep) [reply Druid/DRBasicBlock.class.st:513]: Same as above!

File: Druid/DRBasicBlock.class.st (hunk 3 above)
Comment (guillep) [reply Druid/DRBasicBlock.class.st:521]: Same

Comment (guillep) [inline Druid/DRBasicBlock.class.st:521]:
```suggestion
	"Return a collection of all the blocks that are immediately post dominated by the block given as argument"
```

File: Druid/DRBasicBlock.class.st (hunk 4 above)
Comment (guillep) [reply Druid/DRBasicBlock.class.st:693]: Which methods?

File: Druid/DRCogitSendMarshallHoisting.class.st (hunk 5 above)
Comment (guillep) [reply Druid/DRCogitSendMarshallHoisting.class.st:24]: Yes, thanks for the issue :)

File: Druid/DRDominatorTree.class.st (hunk 7 above)
Comment (guillep) [reply Druid/DRDominatorTree.class.st:240]: Probably not used anymore. I'll add an issue to clean it up

**Pull request merged**